File: Cpp/146.cpp
// Design and implement a data structure for Least Recently Used (LRU) cache. It
// should support the following operations: get and put.
//
// get(key) - Get the value (will always be positive) of the key if the key
// exists in the cache, otherwise return -1. put(key, value) - Set or insert the
// value if the key is not already present. When the cache reached its capacity,
// it should invalidate the least recently used item before inserting a new
// item.
//
// Follow up:
// Could you do both operations in O(1) time complexity?
//
// Example:
//
// LRUCache cache = new LRUCache( 2 /* capacity */ );
//
// cache.put(1, 1);
// cache.put(2, 2);
// cache.get(1);    // returns 1
// cache.put(3, 3); // evicts key 2
// cache.get(2);    // returns -1 (not found)
// cache.put(4, 4); // evicts key 1
// cache.get(1);    // returns -1 (not found)
// cache.get(3);    // returns 3
// cache.get(4);    // returns 4

#include <iostream>
#include <list>
#include <unordered_map>
#include <vector>

using namespace std;

class LRUCache {
public:
    list<pair<int, int>> cache;
    unordered_map<int, list<pair<int, int>>::iterator> seek_ptr;
    size_t _capacity;
    LRUCache(int capacity) { _capacity = capacity; }

    int get(int key) {
        if (!seek_ptr.count(key)) {
            return -1;
        }
        cache.splice(cache.begin(), cache, seek_ptr[key]);
        seek_ptr[key] = cache.begin();
        return cache.front().second;
    }

    void put(int key, int value) {
        if (seek_ptr.count(key)) {
            cache.splice(cache.begin(), cache, seek_ptr[key]);
            seek_ptr[key] = cache.begin();
            cache.front().second = value;
            return;
        }

        if (_capacity == cache.size()) {
            int del_key = cache.back().first;
            cache.pop_back();
            seek_ptr.erase(del_key);
        }
        cache.push_front(make_pair(key, value));
        seek_ptr[key] = cache.begin();
    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

int main() {}

File: Python3/146.py
class Node:
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = self.next = None

class LRUCache:

    def __init__(self, capacity: int):
        self.head = Node(-1, -1)
        self.tail = Node(-1, -1)
        self.head.next = self.tail
        self.tail.prev = self.head
        self.capacity = capacity
        self.cache = {}
    
    def _add_node(self, node):
        second = self.head.next
        self.head.next = node
        node.next, node.prev = second, self.head
        second.prev = node

    def _remove_node(self, node):
        prev, next = node.prev, node.next
        prev.next = next
        next.prev = prev
    
    def _move_to_head(self, node):
        self._remove_node(node)
        self._add_node(node)

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        node = self.cache[key]
        self._move_to_head(node)
        return node.value


    def put(self, key: int, value: int) -> None:
        # update the value
        if key in self.cache:
            node = self.cache[key]
            node.value = value
            self._move_to_head(node)
            return
        
        #add new node
        if len(self.cache) == self.capacity:
            lastnode = self.tail.prev
            self._remove_node(lastnode)
            del self.cache[lastnode.key]
        
        node = Node(key, value)
        self._add_node(node)
        self.cache[key] = node
        return


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
